source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,762 @@
+        break;
+      default:
+        break;
+    }
+  }
+  return modified;
+}
+
+bool FoldSpecConstantOpAndCompositePass::ProcessOpSpecConstantOp(
+    ir::Module::inst_iterator* pos) {
+  ir::Instruction* inst = &**pos;
+  ir::Instruction* folded_inst = nullptr;
+  assert(inst->GetInOperand(0).type ==
+             SPV_OPERAND_TYPE_SPEC_CONSTANT_OP_NUMBER &&
+         "The first in-operand of OpSpecContantOp instruction must be of "
+         "SPV_OPERAND_TYPE_SPEC_CONSTANT_OP_NUMBER type");
+
+  switch (static_cast<SpvOp>(inst->GetSingleWordInOperand(0))) {
+    case SpvOp::SpvOpCompositeExtract:
+      folded_inst = DoCompositeExtract(pos);
+      break;
+    case SpvOp::SpvOpVectorShuffle:
+      folded_inst = DoVectorShuffle(pos);
+      break;
+
+    case SpvOp::SpvOpCompositeInsert:
+      // Current Glslang does not generate code with OpSpecConstantOp
+      // CompositeInsert instruction, so this is not implmented so far.
+      // TODO(qining): Implement CompositeInsert case.
+      return false;
+
+    default:
+      // Component-wise operations.
+      folded_inst = DoComponentWiseOperation(pos);
+      break;
+  }
+  if (!folded_inst) return false;
+
+  // Replace the original constant with the new folded constant, kill the
+  // original constant.
+  uint32_t new_id = folded_inst->result_id();
+  uint32_t old_id = inst->result_id();
+  def_use_mgr_->ReplaceAllUsesWith(old_id, new_id);
+  def_use_mgr_->KillDef(old_id);
+  return true;
+}
+
+ir::Instruction* FoldSpecConstantOpAndCompositePass::DoCompositeExtract(
+    ir::Module::inst_iterator* pos) {
+  ir::Instruction* inst = &**pos;
+  assert(inst->NumInOperands() - 1 == 2 &&
+         "OpSpecConstantOp CompositeExtract requires two non-type non-opcode "
+         "operands.");
+  assert(inst->GetInOperand(1).type == SPV_OPERAND_TYPE_ID &&
+         "The vector operand must have a SPV_OPERAND_TYPE_ID type");
+  assert(
+      inst->GetInOperand(2).type == SPV_OPERAND_TYPE_LITERAL_INTEGER &&
+      "The literal operand must have a SPV_OPERAND_TYPE_LITERAL_INTEGER type");
+
+  analysis::Constant* first_operand_const =
+      FindRecordedConst(inst->GetSingleWordInOperand(1));
+  if (!first_operand_const) return nullptr;
+
+  if (analysis::CompositeConstant* composite_const =
+          first_operand_const->AsCompositeConstant()) {
+    // Case 1: first operand is a non-null composite type constant.
+    uint32_t literal = inst->GetSingleWordInOperand(2);
+    assert(literal < composite_const->GetComponents().size() &&
+           "Literal index out of bound of the composite constant");
+    auto new_const = composite_const->GetComponents().at(literal)->Copy();
+    return BuildInstructionAndAddToModule(std::move(new_const), pos);
+
+  } else if (first_operand_const->AsNullConstant()) {
+    // Case 2: first operand is a constant created with OpConstantNull.
+    return BuildInstructionAndAddToModule(CreateConst(GetType(inst), {}), pos);
```

Comment:
Hmm, the problem is that, there are two sources of the result type.
1) The original instruction. If the module is as valid module, the result type must be the same (or valid to be implicitly converted) with the index'd member type.
2) The index'd member type.

I think I'm using the result type in this pass all the time. As we do have some cases that we convert the type, like
`%cast_to_utint = OpSpecConstantOp %uint IAdd %some_int %int_zero`